src/cregistry/distfile.c

```diff
@@ -0,0 +1,316 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "distfile.h"
+#include "util.h"
+#include "registry.h"
+#include "sql.h"
+
+#include <sqlite3.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Converts a `sqlite3_stmt` into a `reg_distfile`. The first column of the stmt's
+ * row must be the id of a distfile; the second column must be the path of a distfile;
+ * the third either `SQLITE_NULL` or the address of the entry in memory.
+ *
+ * @param [in] userdata sqlite3 database
+ * @param [out] distfile    distfile described by `stmt`
+ * @param [in] stmt     `sqlite3_stmt` with appropriate columns
+ * @param [out] errPtr  unused
+ * @return              true if success; false if failure
+ */
+static int reg_stmt_to_distfile(void* userdata, void** distfile, void* stmt,
+        void* calldata UNUSED, reg_error* errPtr UNUSED) {
+    int is_new;
+    reg_registry* reg = (reg_registry*)userdata;
+    reg_distfile_pk key;
+    Tcl_HashEntry* hash;
+    char* hashkey;
+
+    key.id = sqlite3_column_int64(stmt, 0);
+    key.subdir = strdup((const char*) sqlite3_column_text(stmt, 1));
+    if (!key.subdir) {
+        return 0;
+    }
+    key.path = strdup((const char*) sqlite3_column_text(stmt, 2));
+    if (!key.path) {
+        free(key.subdir);
+        return 0;
+    }
+
+    hashkey = sqlite3_mprintf("%lld:%s", key.id, key.path);
+    if (!hashkey) {
+        free(key.subdir);
+        free(key.path);
+        return 0;
+    }
+    hash = Tcl_CreateHashEntry(&reg->open_distfiles,
+            hashkey, &is_new);
+    sqlite3_free(hashkey);
+
+    if (is_new) {
+        reg_distfile* f = malloc(sizeof(reg_distfile));
+        if (!f) {
+            free(key.subdir);
+            free(key.path);
+            return 0;
+        }
+        f->reg = reg;
+        f->key = key;
+        f->proc = NULL;
+        *distfile = f;
+        Tcl_SetHashValue(hash, f);
+    } else {
+        free(key.subdir);
+        free(key.path);
+        *distfile = Tcl_GetHashValue(hash);
+    }
+    return 1;
+}
+
+/**
+ * Opens an existing distfile in the registry.
+ *
+ * @param [in] reg      registry to open entry in
+ * @param [in] id       port id in the database
+ * @param [in] name     distfile path in the database
+ * @param [out] errPtr  on error, a description of the error that occurs
+ * @return              the distfile if success, NULL if failure
+ */
+reg_distfile* reg_distfile_open(reg_registry* reg, char* id, char* subdir, char* path,
```

Comment:
I'm not sure why I didn't declare the string parameters as const. I don't think we're modifying them in the function, so we might as well do that?